NAVAND-777: add route refresh on demand

dzinad opened the pull request:
### Description

The main idea is that we have 2 types of refresh now: planned (once in refreshInterval) and on-demand (the user triggers the refresh immediately). The planned refresh was present before.
`RouteRefreshController` coordinates these 2 types of refresh. Don't look at the diff, see it as a new file. It changed tremendously. 
`PlannedRefreshController` is responsible for planned refresh.
`ImmediateRefreshController` is responsible for on-demand refresh. 
`RouteRefresher` executes one refresh request.
`RouteRefresherExecutor` makes sure only one refresh request is executed at a time and if the queue is too big, only the current request and the last one from the queue are executed, the intermediate ones are ignored.
`CancellableHandler` is responsible for posting tasks after a timeout and cancelling them on-demand, also notifying the tasks that they were cancelled.
`RouteRefreshStateHolder` is responsible for receiving and keeping route refresh state updates and notifying observers registered via `MapboxNavigation#registerRouteRefreshStateObserver`.


### The previous flow (with only planned refresh):

![planned_route_refresh](https://user-images.githubusercontent.com/32109537/202415810-ed4892f7-c3b8-4b46-af40-f376246505e0.png)

### The new flow (on-demand + planned):

![new_route_refresh](https://user-images.githubusercontent.com/32109537/202418989-e91b4a4f-a2c3-4f20-8c87-9f56eb882d32.png)

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefresherExecutor.kt:
```diff
@@ -0,0 +1,41 @@
+) {
+
+    private val mutex = Mutex()
+    private val queue = ArrayDeque<Pair<List<NavigationRoute>, RouteRefresherProgressCallback>>()
```
Why do we need a queue here? What type of requests overlap and need to be scheduled this way?

dzinad replied:
That's what we discussed: if we invoke `refreshRoutesImmediately` very often, we should:
1) have only 1 running request at a time;
2) execute only current request (if present) and the latest one, not all of them.

LukasPaczos replied:
`refreshRoutesImmediately` has no callback in the current proposal, it's a call that developers make and forget, the call provides no direct side effects (there's no data that developers provide to us and expect us to apply where a coalescing pattern would be needed), and they have no means to track the state and result of it anyway (they don't need an exact ratio of call to callback). This makes me think that we don't need this complexity - if there's already a refresh request in the air when `refreshRoutesImmediately` is called, we can ignore the new call.

dzinad replied:
OK, makes sense.

LukasPaczos commented on libnavigation-core/src/test/java/com/mapbox/navigation/core/routerefresh/PlannedRouteRefreshControllerTest.kt:
```diff
@@ -0,0 +1,645 @@
+        finishRequest(result)
+        startRequest()
+
+        verify(exactly = 0) { stateHolder.onStarted() }
```
I understand the retry shouldn't call start again but here it looks like we didn't call start at all.

dzinad replied:
That's because I did `finishRequest` without `startRequest`. First `startRequest` is not required for the unit test.

LukasPaczos commented on libnavigation-core/src/test/java/com/mapbox/navigation/core/routerefresh/PlannedRouteRefreshControllerTest.kt:
```diff
@@ -0,0 +1,645 @@
+    }
+
+    @Test
+    fun resumeHasRoutesShouldRetry() {
```
If some class called `PlannedRouteRefreshController#resume` without first calling `PlannedRouteRefreshController#pause` we'd have multiple refreshes queued, and they would not respect the general refresh interval. Could we make this more robust so that it's not possible to misuse?

dzinad commented on instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteRefreshStateTest.kt:
```diff
@@ -0,0 +1,547 @@
+
+        delay(2000) // refresh interval + accuracy
+
+        assertEquals(
+            listOf(RouteRefreshExtra.REFRESH_STATE_FINISHED_FAILED),
+            observer.getStatesSnapshot()
+        )
```
It was like this before, I didn't want to break the behaviour. If the routes are invalid (refresh is disabled for all of them), we get (and we used to get) FAILED without STARTED.

dzinad replied:
We're not anymore.

dzinad commented on instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteRefreshStateTest.kt:
```diff
@@ -0,0 +1,547 @@
+    }
+
+    @Test
+    fun routeRefreshIsNotCancelledOnDestroyTest() = sdkTest {
```
It's not about cancelling the job, it's about the observer. I don't expect to receive CANCELLED here because:
a) it was the behaviour before this PR;
b) all observers are removed in `onDestroy` so we would depend on the line order in `onDestroy` which is not cool although manageable.

LukasPaczos replied:
Got it, I think that's fine. I think the test description is just misleading - we are canceling the job, we're just not triggering the observer.

dzinad replied:
Renamed the test.

dzinad commented on instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteRefreshStateTest.kt:
```diff
@@ -0,0 +1,547 @@
+        mapboxNavigation.setNavigationRoutesAndWaitForUpdate(requestedRoutes)
+        mapboxNavigation.refreshRoutesImmediately()
+
+        waitForRefresh()
```
It's the previous behaviour: if routes were changed after 3 failed refresh attempts (and they are changed because we remove expiring data from them), we notify the user. Again, I don't think this behaviour should be broken.

LukasPaczos replied:
Where do the 3 failed request come from in this test though? We're making one on-demand request and expect the failure, why are we waiting for the 3 scheduled requests?

3 request with interval of 10 seconds (as is the test configuration) should exceed our test timeout anyway, so I still don't understand where is this coming from.

dzinad replied:
Ah, I didn't understand the question.

> Why does a failed refresh attempt still invoke RoutesObserver with reason `RoutesExtra.ROUTES_UPDATE_REASON_REFRESH`?

I think it should be like this for on demand refresh. You trigger the request - you get the result. Either success or failure. If it's a failure, we remove expiring data from route -> we get new refreshed routes.

LukasPaczos replied:
> If it's a failure, we remove expiring data from route -> we get new refreshed routes.

I think this is debatable. We're risking clearing all the data on a single failed request whereas the regular cycle requires all 3 requests and their intervals. A small gap in network connectivity (like a request made in a tunnel) would lead to pretty poor user experience.

Alternative would be to only clear the data on 3 unconditional requests but since we reset the counter, if someone calls the on-demand refresh often, we'd never clear the data. This is not ideal but still sounds better to me - clearing the data is a last resort when we're sure the data is already stale, this is not the case after a single failed on-demand refresh.

dzinad replied:
Agreed that it should not be like this. But I think it would be better to use time since last refresh, not the number of attempts. Implemented it in a separate commit. BTW, it also solves the issue with invoking `removeExpiredDataFromRoute` unnecessarily often.

dzinad commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/PlannedRouteRefreshController.kt:
```diff
@@ -0,0 +1,138 @@
+        cancellableHandler.cancelAll()
+    }
+
+    override fun resume() {
+        routesToRefresh?.let {
+            if (retryStrategy.shouldRetry()) {
+                scheduleUpdateRetry(it)
+            }
+        }
+    }
```
It's a bit complicated. I don't like it but I decided not to try to simplify everything perfectly in one step but just to start with something.
Pause will not cause receiving CANCELLED state. Therefore we don't need to dispatch STARTED in resume.
It **was** something like that:
Assume we need 3 attempts to refresh routes. It will be: STARTED -> attempt1 -> attempt2 -> attempt3 -> SUCCESS/FAILED and not STARTED -> attempt1 -> FAILED -> STARTED -> attempt2 -> FAILED -> STARTED -> attempt3 -> SUCCESS/FAILED.
Now `pause` will just cancel attempt2 if it is currently planned and `resume` will say: "resume the attempts starting from attempt2". In order to keep it as it was, `pause` and `resume` do not affect states. 
Does this make sense?

LukasPaczos replied:
I still don't fully understand. `pause` is invoked when an on-demand refresh is executed. If that refresh fails, we'll `resume` without dispatching on-start, so we have:
```
STARTED -> attempt1 -> on-demand (`pause`) -> FAILED -> (`resume` without STARTED) -> SUCCESS/FAILED
```
don't we? We're missing the start for the resumed attempt.

dzinad replied:
Oh, I got confused. `pause` does dispatch the `CANCELLED` state. You are right, `STARTED` should be dispatched here. It was not caught in tests because now every time we do the resume the next moment we actually restart planned route refreshes (because we get new REFRESHED routes, more info here: https://github.com/mapbox/mapbox-navigation-android/pull/6610#discussion_r1032371777) so it doesn't really matter.

dzinad commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshStateChanger.kt:
```diff
@@ -0,0 +1,39 @@
+            RouteRefreshExtra.REFRESH_STATE_CANCELED,
+            null,
+        ),
+        RouteRefreshExtra.REFRESH_STATE_CANCELED to listOf(
+            RouteRefreshExtra.REFRESH_STATE_STARTED,
+            null,
+        ),
```
Yes, it does make sense considering that we can go to FAILED without STARTED.

dzinad commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefresher.kt:
```diff
@@ -0,0 +1,217 @@
+        } else {
+            RouteRefresherResult(
+                false,
+                routes.map { removeExpiringDataFromRoute(it, routeProgressData.legIndex) },
```
Good point, `NAVAND-928`

dzinad commented on instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteRefreshStateTest.kt:
```diff
@@ -0,0 +1,547 @@
+import java.util.concurrent.TimeUnit
+
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
+class RouteRefreshStateTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.java) {
```
Added.

dzinad commented on instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteRefreshStateTest.kt:
```diff
@@ -0,0 +1,547 @@
+import java.util.concurrent.TimeUnit
+
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
+class RouteRefreshStateTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.java) {
```
Added but I used different timeout because I don't want the test to be flaky.

dzinad commented on instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteRefreshOnDemandTest.kt:
```diff
@@ -0,0 +1,216 @@
+            .build()
+        RouteRefreshOptions::class.java.getDeclaredField("intervalMillis").apply {
+            isAccessible = true
+            set(routeRefreshOptions, 10_000L)
```
Thus we would increase the probability of a flaky test which I definitely don't want to do.

dzinad commented on instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteRefreshStateTest.kt:
```diff
@@ -0,0 +1,547 @@
+        val requestedRoutes = requestRoutes(enableRefresh = false)
+        mapboxNavigation.setNavigationRoutes(requestedRoutes)
+
+        delay(7000) // refresh interval + accuracy
```
I don't think it can be tested properly in unit tests. They test many parts of the SDK along with the server interaction. While writing these tests I've discovered a couple of bugs and inconsistencies. What's more, they give you a feel of how the SDK works which might not have been so obvious before writing/reading them. I find them very useful and would like to keep them.
What exactly is the problem with additional 2 mins? Aren't the tests paralleled somehow?

LukasPaczos replied:
This is 2 minutes for a single test class. As we keep expanding at some point these types of granular addition will just unbearably extend the CI execution time. We're not at the critical point for the instrumentation tests job yet but we should be mindful of that as we go. If we can trim the execution down (for example, 7 seconds of a delay really shouldn't be necessary), let's do that.

> Aren't the tests paralleled somehow?

No, we're not doing [test sharding](https://firebase.google.com/docs/test-lab/android/instrumentation-test#sharding) yet but that idea has been floated in the past. Something to consider cc @SevaZhukov @dchernousov.

dzinad replied:
I just mean that we shouldn't be held back from writing tests because they require some time to execute... In order not to be flaky at the same time. I don't know how to reduce the delay here:
1) I want to check states triggered by refresh on demand;
2) in order to do that I have to trigger refresh in-between planned attempts;
3) I have to wait for the first planned attempt;

If I make the interval 2-3 seconds instead of 5, code execution time will play a much bigger role in this test being stable.
In general, it's a common problem for tests that check that something is done with correct interval. And I think we should have such tests. For example, it would check that we don't do by mistake any unnecessary requests and so on.

LukasPaczos replied:
> I just mean that we shouldn't be held back from writing tests because they require some time to execute...

Of course not, test coverage always prevails over test execution time. I just want us to be mindful, that's all. If you think that we can't optimize execution time any further without impacting stability, then let's leave this as is.

dzinad replied:
I redid some of the tests into Robolectric tests.

dzinad commented on instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteRefreshStateTest.kt:
```diff
@@ -0,0 +1,547 @@
+import java.util.concurrent.TimeUnit
+
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
+class RouteRefreshStateTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.java) {
```
I don't want to introduce flakiness. It's not a unit test where you can advance the dispatcher with 1 ms precision. Auxiliary code execution takes time that we should account for.
I prefer the stability.
Unit-tests discussion here: https://github.com/mapbox/mapbox-navigation-android/pull/6610#discussion_r1031450251

VysotskiVadim commented on instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteRefreshOnDemandTest.kt:
```diff
@@ -0,0 +1,286 @@
+        failedRefreshHandlerWrapper.failResponse = false
+    }
+
+    @Test(timeout = 10_000)
```
`sdkTest` fails with timeout of 30 seconds by default and it can be configured. TBH I'm not sure which way of configuring timeout is better 🙂

VysotskiVadim reviewed (commented):
I'm very biased about route refresh system as I'm the one who migrated it to coroutines. I have a suggestion of how to achieve the same result writing less code. See the diff. @dzinad , do you see any refresh cases which the diff doesn't cover?

I must admit that new route refresh system may be easier and more comfortable for not biased developers(not the ones who implemented new or current route refresh system).
@LukasPaczos, @RingerJK, which way do you think should we go, modifying existing coroutine based system or switching to the new presented in this PR?
```
Index: libnavigation-core/src/test/java/com/mapbox/navigation/core/routerefresh/RouteRefreshControllerTest.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/libnavigation-core/src/test/java/com/mapbox/navigation/core/routerefresh/RouteRefreshControllerTest.kt b/libnavigation-core/src/test/java/com/mapbox/navigation/core/routerefresh/RouteRefreshControllerTest.kt
--- a/libnavigation-core/src/test/java/com/mapbox/navigation/core/routerefresh/RouteRefreshControllerTest.kt	(revision 592818bbba6bf65accf7a4082311e94ebf938efe)
+++ b/libnavigation-core/src/test/java/com/mapbox/navigation/core/routerefresh/RouteRefreshControllerTest.kt	(date 1669824835092)
@@ -129,6 +129,27 @@
             }
         }
 
+    @Test
+    fun `route refreshes immediately`() =
+        runBlockingTest {
+            val (initialRoute, refreshedRoute) = createTestInitialAndRefreshedTestRoutes()
+            val routeRefreshStub = RouteRefreshStub().apply {
+                setRefreshedRoute(refreshedRoute)
+            }
+            val routeRefreshController = createRouteRefreshController(
+                routeRefresh = routeRefreshStub,
+                routeRefreshOptions = RouteRefreshOptions.Builder()
+                    .intervalMillis(30_000)
+                    .build(),
+            )
+
+            val refreshJob = async { routeRefreshController.refresh(listOf(initialRoute)) }
+            assertTrue(refreshJob.isActive)
+            routeRefreshController.refreshImmediately()
+            assertEquals(listOf(refreshedRoute), refreshJob.getCompletedTest().routes)
+            refreshJob.cancel()
+        }
+
     @Test
     fun `should refresh route with any annotation`() = runBlockingTest {
         val routeWithoutAnnotations = createNavigationRoute(
Index: libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt b/libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt
--- a/libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt	(revision 592818bbba6bf65accf7a4082311e94ebf938efe)
+++ b/libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt	(date 1669825162219)
@@ -15,14 +15,21 @@
 import com.mapbox.navigation.core.RouteProgressDataProvider
 import com.mapbox.navigation.core.directions.session.RouteRefresh
 import com.mapbox.navigation.core.ev.EVRefreshDataProvider
+import com.mapbox.navigation.utils.internal.logD
 import com.mapbox.navigation.utils.internal.logE
 import com.mapbox.navigation.utils.internal.logI
+import com.mapbox.navigation.utils.internal.logW
 import kotlinx.coroutines.CancellationException
 import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.Deferred
 import kotlinx.coroutines.async
 import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.cancelChildren
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.Channel.Factory.RENDEZVOUS
 import kotlinx.coroutines.coroutineScope
 import kotlinx.coroutines.delay
+import kotlinx.coroutines.selects.select
 import kotlinx.coroutines.suspendCancellableCoroutine
 import kotlinx.coroutines.withTimeoutOrNull
 import java.util.Date
@@ -43,6 +50,8 @@
     private val localDateProvider: () -> Date,
 ) {
 
+    private val immediateRefreshRequestsChannel = Channel<Unit>(RENDEZVOUS)
+
     private var state: RouteRefreshStateResult? = null
         set(value) {
             if (field == value) return
@@ -92,6 +101,12 @@
         }
     }
 
+    fun refreshImmediately() {
+        if (immediateRefreshRequestsChannel.trySend(Unit).isFailure) {
+            logW { "immediate refresh is ignored as route refresh is in progress already" }
+        }
+    }
+
     fun registerRouteRefreshStateObserver(observer: RouteRefreshStatesObserver) {
         observers.add(observer)
         state?.let { observer.onNewState(it) }
@@ -150,7 +165,8 @@
         var timeUntilNextAttempt = async { delay(routeRefreshOptions.intervalMillis) }
         try {
             repeat(FAILED_ATTEMPTS_TO_INVALIDATE_EXPIRING_DATA) {
-                timeUntilNextAttempt.await()
+                waitForIntervalOrImmediateRefreshSignal(timeUntilNextAttempt)
+
                 if (it == 0) {
                     onNewState(RouteRefreshExtra.REFRESH_STATE_STARTED)
                 }
@@ -180,6 +196,25 @@
         )
     }
 
+    // inspired by https://github.com/Kotlin/kotlinx.coroutines/issues/2867#issuecomment-998553524
+    private suspend fun waitForIntervalOrImmediateRefreshSignal(
+        timeUntilNextAttempt: Deferred<Unit>
+    ) {
+        coroutineScope {
+            select<Unit> {
+                timeUntilNextAttempt.onAwait {
+                    logD { "starting refresh by interval" }
+                }
+                immediateRefreshRequestsChannel.onReceive {
+                    logD { "starting immediate refresh" }
+                }
+            }.also {
+                timeUntilNextAttempt.cancel()
+                coroutineContext.cancelChildren()
+            }
+        }
+    }
+
     private fun removeExpiringDataFromRoute(
         route: NavigationRoute,
         currentLegIndex: Int,
@@ -344,4 +379,4 @@
         data class Success(val route: NavigationRoute) : RouteRefreshResult()
         data class Fail(val error: NavigationRouterRefreshError) : RouteRefreshResult()
     }
-}
+}
\ No newline at end of file

```

dzinad commented:
Just rebasing, don't pay attention ^

dzinad commented:
@VysotskiVadim 
I've run my tests with your code and the following ones fail:
`route_refresh_on_demand_executes_before_refresh_interval`,
`routeRefreshOnDemandBetweenPlannedAttemptsTest`,
`routeRefreshOnDemandFailsThenPlannedTest`,
`routeRefreshOnDemandForInvalidRoutes`,
`successfulRouteRefreshOnDemandTest`,
`failedRouteRefreshOnDemandTest`,
`routeRefreshOnDemandFailsBetweenPlannedAttemptsTest`

In general, Im not against using coroutines here. But I found `RouteRefreshContoller` class too heavy and definitely didn't want to make it grow even more. And still strongly don't want to. I tried to break it up into several classes making them follow if not single-responsibility, then at least smaller-number-reponsibility pattern. And with that approach I found it easier to use callbacks. If you have any idea how to make that code more synchronous using coroutines, I'm all in. But I would be against returning to 1 class `RouteRefreshController` that handles everything refresh-related.

VysotskiVadim commented:
@dzinad , 
I don't see any violations of SRP in old `RouteRefreshController` 🙂 
As for failed tests I believe the old coroutines based system can be adopted for those requirements. I'm not sure if all the requirements verified in tests are strict, maybe some of them debatable. I'm also not sure if it makes sense to put some effort in old system to adopt just to prove my point before we decide the way we want to go.

I'm not agains your implementation. But I'm a biased author of the old coroutines-based implementation and for me it seems so much simpler and understandable. Code is always more understandable for the authors as the code represents the way authors see their features. You see route refresh a different way, and I'm sure you find your implementation to be much understandable then the old one. That's why it's very interesting for me which way @LukasPaczos  and @RingerJK find the more simple and understandable. I think their opinion won't be as biased as mine.

dzinad commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/ImmediateRouteRefreshController.kt:
```diff
@@ -0,0 +1,38 @@
+                stateHolder.onSuccess()
+            } else {
+                stateHolder.onFailure(null)
+                plannedRefreshController.resume()
```
Because success result means new refreshed routes -> we don't have to resume refreshing the old ones. We'll get a notification with reason = REFRESH and relaunch planned refresh.

dzinad commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshStateHolder.kt:
```diff
@@ -0,0 +1,61 @@
+        message: String? = null
+    ) {
+        val oldState = this.state?.state
+        if (oldState != state && RouteRefreshStateChanger.canChange(from = oldState, to = state)) {
```
Yes, it can. 
It's just the generalization of what we have on main. But there we only switch to CANCELED if we are in STARTED state. Here I introduced the table for all the states.

dzinad commented:
BTW, there was a request from our customer to add an API that pauses/resumes refreshes. With this code base, it's super easy to do. So I'd be in favour of keeping the refactored code.

dzinad commented:
@LukasPaczos I moved experimental API to another interface (see changelog for details). Since we want to add more refresh APIs (also pause/resume quite soon), I thought it would be nice to extract it into a separate interface in order not to keep overloading `MapboxNavigation`. Just wanted to double-check it's OK for experimental APIs.

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/ImmediateRouteRefreshController.kt:
```diff
@@ -0,0 +1,52 @@
+        routes: List<NavigationRoute>,
+        callback: (RouteRefresherResult) -> Unit
+    ) {
+        if (routes.isEmpty()) {
+            return
+        }
```
We should either throw or notify the callback with a failure.

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/PlannedRouteRefreshController.kt:
```diff
@@ -0,0 +1,148 @@
+        cancellableHandler.cancelAll()
+        routesToRefresh = null
+        if (routes.isEmpty()) {
+            logI("Routes are empty", RouteRefreshLog.LOG_CATEGORY)
```
```suggestion
            logI("Routes are empty, nothing to refresh", RouteRefreshLog.LOG_CATEGORY)
```
nit

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/PlannedRouteRefreshController.kt:
```diff
@@ -0,0 +1,148 @@
+                )
+            val logMessage = "No routes which could be refreshed. $message"
+            logI(logMessage, RouteRefreshLog.LOG_CATEGORY)
+            stateHolder.onFailure(logMessage)
```
```suggestion
            stateHolder.onStarted()
            stateHolder.onFailure(logMessage)
```
wondering if this wouldn't be better to keep the state transition more predictable

dzinad replied:
That's the previous behaviour. Didn't want to break it.
Discussing this in another thread.

dzinad replied:
OK, that's not related.
It doesn't look worse and it's experimental. Why not.

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshControllerImpl.kt:
```diff
@@ -0,0 +1,57 @@
+
+    override fun requestImmediateRouteRefresh() {
+        val routes = plannedRouteRefreshController.routesToRefresh
+        if (routes.isNullOrEmpty()) {
+            return
+        }
```
```suggestion
        if (routes.isNullOrEmpty()) {
            log that there are no routes to be refreshed
            return
        }
```

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/PlannedRouteRefreshController.kt:
```diff
@@ -0,0 +1,148 @@
+        }
+    }
+
+    private fun scheduleUpdateRetry(routes: List<NavigationRoute>, shouldNotifyOnStart: Boolean) {
+        postAttempt { executePlannedRefresh(routes, shouldNotifyOnStart = shouldNotifyOnStart) }
+    }
```
```suggestion
    private fun scheduleUpdateRetry(routes: List<NavigationRoute>, shouldNotifyOnStart: Boolean) {                
        retryStrategy.onNextAttempt()
        postAttempt { executePlannedRefresh(routes, shouldNotifyOnStart = shouldNotifyOnStart) }
    }
```
this looks clearer to me

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/PlannedRouteRefreshController.kt:
```diff
@@ -0,0 +1,148 @@
+            }
+
+            override fun onResult(routeRefresherResult: RouteRefresherResult) {
+                retryStrategy.onNextAttempt()
```
```suggestion
```
as mentioned above

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RetryRouteRefreshStrategy.kt:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    fun shouldRetry(): Boolean {
+        return attemptNumber <= maxRetryCount
```
```suggestion
        return attemptNumber < maxRetryCount
```
as mentioned above

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefresherExecutor.kt:
```diff
@@ -0,0 +1,36 @@
+    private val timeout: Long,
+) {
+
+    private var hasCurrentRequest = false
+
+    fun postRoutesToRefresh(
+        routes: List<NavigationRoute>,
+        callback: RouteRefresherProgressCallback
+    ) {
+        scope.launch {
+            if (!hasCurrentRequest) {
+                hasCurrentRequest = true
+                callback.onStarted()
+                val result = routeRefresher.refresh(routes, timeout)
+                callback.onResult(result)
+                hasCurrentRequest = false
+            }
+        }
+    }
+}
```
I see a couple of problems with this codebase.

First that it's dangerous to leave callbacks without a response. If a piece of code provides a callback, it expects to be called back at some point to resume operations. Now, if the flag is set the callback is ignored.

Second, is this scenario possible?
- Start planned refresh.
- Wait for the interval to elapse.
- While the planned request is in flight, request immediate refresh.
- The immediate refresh will cancel and pause the planned refresh but the  `hasCurrentRequest` flag will never be reset, so the immediate request will be ignored and we end up with a  "deadlock".

Am I missing something?

dzinad replied:
> First that it's dangerous to leave callbacks without a response. If a piece of code provides a callback, it expects to be called back at some point to resume operations. Now, if the flag is set the callback is ignored.

Agreed. I'll fix that.

> Second, is this scenario possible?

It is. Thanks. I'll fix and add a test.

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshControllerProvider.kt:
```diff
@@ -1,28 +1,70 @@
+            refreshObserversManager,
+            ExpiringDataRemover { Date() },
+            Time.SystemImpl,
+            routeRefreshOptions.intervalMillis * (PlannedRouteRefreshController.MAX_RETRY_COUNT + 1)
```
```suggestion
            routeRefreshOptions.intervalMillis * 3
```
The process for clearing out-of-date information is now based purely on time, it's only artificially coupled with number of attempts because there could be many more attempts happening with immediate refresh being taken into account.

Could we remove the concept of attempts from the codebase in this case? This would in some way simplify this already complex setup and the only side effect is that a `start/cancel/success/failure` `RouteRefreshStatesObserver` callbacks would be invoke for each request instead of being invoked every third request.

dzinad replied:
> and the only side effect is that a start/cancel/success/failure RouteRefreshStatesObserver callbacks would be invoked for each request instead of being invoked every third request.

I was trying to avoid breaking this behaviour. IMO this is a breaking change.

LukasPaczos replied:
The retry logic is an entirely internal construct, we don't specify that behavior anywhere in the docs or the public API. The only entry in public, stable `RouteRefreshOptions` docs is:
```
In case of a route refresh failing for `3 * `[intervalMillis] (default is 15 min), expired incidents and congestion annotations will be removed from the route...
```

On top of that, the current `REFRESH_STATE_FINISHED_FAILED` doc says:
```
The state becomes [REFRESH_STATE_FINISHED_FAILED] when a route refresh failed and the route is cleaned up of expired data, see [RouteRefreshOptions] for details.
```
which is not true for this PR because a `REFRESH_STATE_FINISHED_FAILED` will also be delivered for a failed immediate request which might not have cleaned up expired data yet.

I think we should consider refactoring the `REFRESH_STATE_FINISHED_FAILED` doc (since it's experimental) to delete the expired data removal language, and expose a new `REFRESH_STATE_CLEARED_EXPIRED` to be deliver after a failure that also exceeded the time threshold (or some other alternative to this).

dzinad replied:
Actually, if I just change `PlannedRouteRefreshController.MAX_RETRY_COUNT + 1` to 3, route refresh state callbacks behaviour won't change.

dzinad replied:
> I think we should consider refactoring the REFRESH_STATE_FINISHED_FAILED doc (since it's experimental) to delete the expired data removal language, and expose a new REFRESH_STATE_CLEARED_EXPIRED to be deliver after a failure that also exceeded the time threshold (or some other alternative to this).

I like that!

dzinad commented on libnavigation-core/api/current.txt:
```diff
@@ -26,6 +26,7 @@ package com.mapbox.navigation.core {
     method public com.mapbox.navigation.core.reroute.NavigationRerouteController? getRerouteController();
     method public com.mapbox.navigation.core.trip.session.eh.RoadObjectMatcher getRoadObjectMatcher();
     method public com.mapbox.navigation.core.trip.session.eh.RoadObjectsStore getRoadObjectsStore();
+    method public com.mapbox.navigation.core.routerefresh.RouteRefreshController getRouteRefreshController();
```
That's weird. It's marked as experimental...

dzinad replied:
Looks like a metalava issue. It didn't mark `MapboxNavigation#experimental` property with the annotation either. I could convert it into a method but I'm not sure if it's worth it. After all, the interface itself is experimental so opt-in is required anyway.

LukasPaczos replied:
:+1: let's leave it as is then.

dzinad commented:
@LukasPaczos I addressed your comments.
BTW, adding `REFRESH_STATE_CLEARED_EXPIRED` and controlling it was super easy after refactoring. So I think it did make sense to refactor at least partially.

VysotskiVadim commented on changelog/unreleased/features/6610.md:
```diff
@@ -0,0 +1,18 @@
+- Added `RouteRefreshController` interface to manage route refreshes. Retrieve it via `MapboxNavigation#routeRefreshController`.
```
Does exposing `RouteRefreshController` makes our SDK easier to harder to change?
I found myself in a complex situations with reroute logic a few times. It was very hard to change its structure, because we exposed it to the users in our API. Maybe exposing `MapboxNavigation#refreshRoute()` leaves us more flexibility for future changes? 🤔

dzinad replied:
It's just an interface that accumulates refresh-related API.

Zayankovsky commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt:
```diff
@@ -1,360 +1,38 @@
-        data class Fail(val error: NavigationRouterRefreshError) : RouteRefreshResult()
-    }
+@ExperimentalPreviewMapboxNavigationAPI
+interface RouteRefreshController {
```
Adding new functions to an `interface` is a breaking change. Are you sure we won't add any after the interface is no longer experimental?

dzinad replied:
Oh shoot, you're right.
I didn't design this interface to be implemented by customer. So it doesn't look like a pure breaking change to me, although technically it is.
It can be easily fixed by just exposing a class, only marking some of the methods public. But I'd like to know your opinion whether it's really an issue. If the interface is not supposed to be implemented, should adding new methods to it be considered a breaking change? I just don't want us to be artificially limiting ourselves. Exposing some API through an interface instead of a class is generally a good approach, but we wouldn't be able to use it if we can't add new methods. cc @mapbox/navigation-android

Zayankovsky replied:
What is wrong with using a class? Also we can consider making the interface sealed, this way Kotlin developers will not be able to implement it.

dzinad replied:
There is nothing wrong. It's just that hiding an implementation behind an interface is good because it doesn't expose any unnecessary details.
I'm just trying to look from the fresh perspective. Is it really a breaking change? Can we use interfaces as API or not?
Sealed interface only solves the problem for kotlin. Implementing it from Java is allowed: there'll be a warning, but it'll compile.

dzinad replied:
Alright, I exposed a class instead.

dzinad commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefresherResultProcessor.kt:
```diff
@@ -0,0 +1,43 @@
+                    result.refreshedRoutesData
+                )
+                stateHolder.onClearedExpired()
+                if (result.refreshedRoutesData != newRoutesData) {
```
@VysotskiVadim I can't move this check to ObserversManager because here I compare the original data with the data we got after removing the expired part. I don't think that the ObserversManager should know anything about that.

VysotskiVadim commented on instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteRefreshStateTest.kt:
```diff
@@ -0,0 +1,307 @@
+import java.util.concurrent.TimeUnit
+
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
+class RouteRefreshStateTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.java) {
```
I noticed that scenarios of `RouteRefreshStateTest` and `RouteRefreshOnDemandTest` are very similar. Is it possible to move state checks to other instrumentation tests?

dzinad replied:
Yes, well, that's what I wanted to avoid.
These states are experimental and subject to change. Idk why someone even wants to use them, but whatever.
And I don't want to break tests that check something else refresh-related because we decided states should work differently.

dzinad replied:
Discussed internally.

VysotskiVadim commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/CancellableHandler.kt:
```diff
@@ -0,0 +1,34 @@
+    private val scope: CoroutineScope
+) {
+
+    private val jobs = linkedSetOf<Job>()
```
when could it have many jobs? I guess we always refresh the same set of routes, don't we?

dzinad replied:
Well, I think right now we really can't get multiple jobs here, but this class doesn't know anything about how it's going to be used. For example, it shouldn't know that there is some executor that will filter out the request if we already have a running request.

dzinad replied:
Removed this class after all.

VysotskiVadim commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/PlannedRouteRefreshController.kt:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    private fun postAttempt(attemptBlock: suspend () -> Unit) {
+        cancellableHandler.postDelayed(
+            timeout = routeRefreshOptions.intervalMillis,
+            block = attemptBlock,
+            cancellationCallback = { stateHolder.onCancel() }
+        )
```
Note: lets discuss cancelable handler vs routes scope.

dzinad replied:
Discussed internally.

VysotskiVadim commented on libnavigation-core/src/test/java/com/mapbox/navigation/core/routerefresh/ExpiringDataRemoverTest.kt:
```diff
@@ -0,0 +1,192 @@
+
+    @Test
+    fun removeExpiringDataFromRoutes() {
+        every {
+            localDateProvider()
+        } returns parseISO8601DateToLocalTimeOrNull("2022-06-30T20:00:00Z")!!
```
it's probably more of the matter or preferences though

dzinad replied:
Well, here I set the time it should return in the test (because the test holds the context of what time we should return), while the object is created earlier. Since there is only one test, I can move the object creation into the test, but I'm not sure if it will make a lot of sense. If I add another test in the future, I may want different time.

dzinad commented:
@VysotskiVadim I'm gonna merge it into alpha. If you find there is something important I haven't fixed here, I can always follow-up in a separate PR.
